openvdb/points/PointScatter.h

```diff
@@ -0,0 +1,676 @@
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @author Nick Avramoussis
+///
+/// @file PointScatter.h
+///
+/// @brief Various point scattering methods for generating VDB Points.
+///
+///  Note that this architecture is heavily based off the similar
+///  functionality for more generic position scattering in
+///  tools/PointScatter.h, with an additional set of free functions for
+///  convenience.
+///
+///  All random number calls are made to the same generator to produce
+///  temporarily consistent results in relation to the provided seed. This
+///  comes with some multi-threaded performance trade-offs.
+///
+#ifndef OPENVDB_POINTS_POINT_SCATTER_HAS_BEEN_INCLUDED
+#define OPENVDB_POINTS_POINT_SCATTER_HAS_BEEN_INCLUDED
+
+#include <algorithm>
+#include <thread>
+
+#include <openvdb/openvdb.h>
+#include <openvdb/tree/LeafManager.h>
+#include <openvdb/tools/Prune.h>
+#include <openvdb/util/NullInterrupter.h>
+
+#include "PointDataGrid.h"
+#include "AttributeArray.h"
+#include "PointCount.h"
+
+#include <tbb/parallel_sort.h>
+#include <tbb/parallel_for.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace points {
+
+/// @brief Uniformly scatter a total amount of points in active regions
+///
+/// @param grid   A source grid. The resulting PointDataGrid will copy this grids
+///               transform and scatter in its active voxelized topology. Topology
+///               will be deactivated in areas where zero points are scattered.
+/// @param seed   A seed for the RandGenT
+/// @param count  The total number of points to scatter
+/// @note returns the scattered PointDataGrid
+template<
+    typename RandGenT,
+    typename GridT,
+    typename PointDataGridT = points::PointDataGrid>
+inline typename PointDataGridT::Ptr
+pointScatter(const GridT& grid,
+             const unsigned int seed,
+             const Index64 count);
+
+/// @brief Uniformly scatter a fixed number of points per active voxel. If the pointsPerVoxel
+///        value provided is a fractional value, each voxel calculates a delta value of
+///        how likely it is to contain an extra point.
+///
+/// @param grid            A source grid. The resulting PointDataGrid will copy this grids
+///                        transform and scatter in its active voxelized topology. Topology
+///                        will be deactivated in areas where zero points are scattered (if
+///                        pointsPerVoxel is between 0.0f and 1.0f)
+/// @param seed            A seed for the RandGenT
+/// @param pointsPerVoxel  The number of points to scatter per voxel. Must not be less than 0
```

Comment:
I wouldn't say "must not be less than 0" - just make sure it errors in this case. It think it should be expected that a negative points-per-voxel value is invalid input.

Reply:
sure